Alexander/react build

n0mn0m opened the pull request:
Setting this up as a draft for now. This adds serving the new react frontend to the appstore. 

- Django app `frontend` added to serve the helx-ui assets
- Setup multistage build to copy the build artifacts from helx-ui
- Add `bin/appstore` command to pull frontend assets into the local environment
- Add docs to `frontend/views.py` describing how this works
- Update README

This builds off the `dj3` and `drf` branches, the majority of changes are to `frontend/` and `appstore/settings/base.py`/`appstore/urls.py`.

helxplatform/helx-ui#10 needs to merge first for a good image flow to be available from the ui project.

Commit:
Begin work on integrating react frontend

Setup CORS when in dev mode so the frontend can make calls without being served from Django
Add init mvp for loading react content

Commit:
Add Django debug toolbar

Assist in CORS debugging, and helpful in general

Commit:
Add CSRF origins when in debug local dev.

Commit:
Update frontend serving

Setup django app to serve the helx-ui assets at /frontend

- Update Django app name and path
- Add static to templates for loading the frontend from static
- Detail configuration in `frontend/views.py`
- Add helx-ui to Dockerfile using multistage build to bring in artifacts
- Update README
- Add bin/appstore command to fetch frontend assets for local development

n0mn0m commented:
helxplatform/helx-ui#10

@connectbo @mbwatson this is the `appstore` piece of the puzzle. Let me know what feedback you have.

n0mn0m commented:
Currently route conflicts can reroute the user to the existing Django template pages. For instance after going to `frontend` if react attempts to route to `apps/` this will conflict with the existing Django `apps/` route. Since that route has priority in `urls.py` the user will end up in the existing frontend app. We need to look into isolating these routes while we serve both frontends.

mbwatson commented:
we don't have to use the route `/apps` on the frontend. maybe `applications` or `appstore`

n0mn0m commented:
That works for me. I just wanted to call out the potential for that. Our current routes are [here](https://github.com/helxplatform/appstore/blob/alexander/react-build/appstore/appstore/urls.py). I also wanted to write this down in case anybody had an idea or approach that might make this a non issue across the board.

n0mn0m commented:
@mbwatson @connectbo in the `Dockerfile` you will need to reference a `helx-ui` built that has the updates in helxplatform/helx-ui#10 since that has the path updates for react. Otherwise paths won't resolve. Bit of a chicken and egg for testing till we know we have the `ui` stuff ready to merge.

Commit:
Correct url path in frontend

Commit:
Update image run script

- Added volume mounts so that bin/appstore image run works locally
- Update help output

Main goal was to make running the local docker image work which requires mounting paths for tycho to make use of. This can be used particularly in stub mode before pushing the image for use in a helm chart.

ismagilani commented on appstore/README.md:
```diff
@@ -1,86 +1,100 @@
+
 Appstore is packaged as a Docker image. It is a non-root container, meaning the user is not a superuser. It packages
-a branch of Tycho cloned within the appstore hierarchy.
+a frontend copied from [helx-ui](https://hub.docker.com/repository/docker/helxplatform/helx-ui).
```
I would include the tycho line as well since AppStore now includes the tycho library.

n0mn0m replied:
My main thought on that is it comes from `pypi` and is `pip` installed so we don't vendor it the same way, but I can add a note back for it.

ismagilani commented on appstore/core/templates/account/login.html:
```diff
@@ -1,4 +1,4 @@
-{% extends "account/base.html" %}
+{% extends "base.html" %}
```
Was templates/accounts a directory created by us at some point and you are just renaming and restructuring for better?

n0mn0m replied:
This was part of the Django 3 PR. There was an update to the `allauth` project that changed our references.

n0mn0m replied:
https://github.com/pennersr/django-allauth/blob/master/ChangeLog.rst#backwards-incompatible-changes-1

ismagilani commented on appstore/appstore/settings/base.py:
```diff
@@ -67,6 +81,24 @@
     'middleware.session_idle_timeout.SessionIdleTimeout'
 ]
 
+if DEBUG=="True" and DEV_PHASE in ("local", "stub", "dev"):
```
@cnbennett3 would this work given the boolean change you did with DEBUG to make it work earlier?

n0mn0m replied:
I'll test again with all the recent merges and make sure this still works/update for the change.

Commit:
Update debug settings
